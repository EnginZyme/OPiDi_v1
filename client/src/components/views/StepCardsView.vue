<template>
  <draggable
    class="step_cards_view__step_cards_view"
    v-bind="dragOptions"
    :list="list"
    :value="value"
    tag="div"
    @start="dragging = true"
    @end="dragging = false"
    @input="emitter"
  >
    <div v-for="step in realValue" :key="step.id">
      <div
        v-if="step.type === 'loop'"
        class="step_cards_view__step_card"
        :class="stepID === step.id ? 'step_cards_view__step_card--chosen' : ''"
        @click="
          stepID = step.id;
          formType = step.type;
        "
        v-badge="step.num_iterations"
      >
        <p class="step_cards_view__step_card__main_text">{{ step.name }}</p>
        <p class="step_cards_view__step_card__sub_text">
          {{ getFormTitleFromStepType(step.type) }}
        </p>
      </div>
      <div
        v-else
        class="step_cards_view__step_card"
        :class="stepID === step.id ? 'step_cards_view__step_card--chosen' : ''"
        @click="
          stepID = step.id;
          formType = step.type;
        "
      >
        <p class="step_cards_view__step_card__main_text">{{ step.name }}</p>
        <p class="step_cards_view__step_card__sub_text">
          {{ getFormTitleFromStepType(step.type) }}
        </p>
      </div>
      <slot v-if="step.type === 'loop'">
        <StepCardsView
          :list="step.substeps"
          class="step_cards_view__step_card__sub_step_cards_view"
        />
      </slot>
    </div>
  </draggable>
</template>

<script>
import draggable from "vuedraggable";
import types from "../../types";
import { mapGetters } from "vuex";

/**
 * A component that renders the step cards view.
 */
export default {
  name: "StepCardsView",
  components: {
    draggable,
  },
  computed: {
    ...mapGetters(["protocolEditorVariables"]),
    /**
     * A read/write object for the currently selected step.
     */
    stepID: {
      get: function() {
        return this.protocolEditorVariables.stepsstepID;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_STEP_VARIABLES, {
          stepsstepID: newValue,
        });
      },
    },
    /**
     * A read/write object for the form type of the currently selected step.
     */
    formType: {
      get: function() {
        return this.protocolEditorVariables.stepsFormType;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_STEP_VARIABLES, {
          stepsFormType: newValue,
        });
      },
    },
    dragOptions() {
      return {
        animation: 0,
        group: "description",
        disabled: false,
        ghostClass: "ghost",
      };
    },
    // this.value when input = v-model
    // this.list  when input != v-model
    realValue() {
      return this.value ? this.value : this.list;
    },
  },

  methods: {
    emitter(value) {
      this.$emit("input", value);
    },
    /**
     * Returns the corresponding form title for a step type.
     *
     * @param {string} type - The said step type.
     */
    getFormTitleFromStepType(type) {
      if (type === "array_transfer") return "Array Transfer";
      if (type === "simple_transfer") return "Simple Transfer";
      if (type === "sequence_transfer") return "Sequence Transfer";
      if (type === "loop") return "Loop";
      if (type === "bioshake_3000t") return "Bioshake 3000T";
      if (type === "timer") return "Timer";
      if (type === "pause") return "Pause";
      if (type === "slack_message") return "Slack Message";
    },
  },
  props: {
    value: {
      required: false,
      type: Array,
      default: null,
    },
    list: {
      required: false,
      type: Array,
      default: null,
    },
  },
};
</script>

<style scoped>
@import url("https://fonts.googleapis.com/css2?family=Rubik:wght@300&display=swap");

.step_cards_view__step_card__sub_step_cards_view {
  padding: 20px;
  margin-bottom: 10px;
}
.step_cards_view__step_card {
  width: 310px;
  height: 48px;
  display: grid;
  grid-template-rows: 12px 10px;
  grid-row-gap: 5px;
  background: #efefff;
  box-shadow: 2px 2px 9px 2px rgba(0, 0, 0, 0.25);
  border-radius: 3px;
  padding-left: 33px;
  padding-right: 16px;
  padding-top: 14px;
  padding-bottom: 2px;
  margin-left: 10px;
  cursor: pointer;
}
.step_cards_view__step_cards_view {
  display: grid;
  grid-auto-rows: minmax(40px, auto);
  grid-row-gap: 20px;
}
.step_cards_view__step_card--chosen {
  box-shadow: 4px 4px 4px 2px rgba(255, 255, 255, 0.25),
    2px 2px 9px 2px #2196f3aa;
}
.step_cards_view__step_card:active {
  border: thin solid #2196f3aa;
}
.step_cards_view__step_card__main_text {
  width: 232px;
  height: 16px;
  font-size: 12px;
  color: #040716;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-family: "Rubik", sans-serif;
  font-weight: 600;
  margin: 0px;
}
.step_cards_view__step_card__sub_text {
  font-size: 10px;
  color: #767676;
  justify-self: end;
  margin: 0px;
}
</style>
