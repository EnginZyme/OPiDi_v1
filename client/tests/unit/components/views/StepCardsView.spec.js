import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import StepCardsView from "@/components/views/StepCardsView";

const localVue = createLocalVue();

localVue.use(Vuex);

const sampleList = null;
const sampleValue = [
  {
    type: "sequence_transfer",
    name: "Sequence Transfer (0)",
    id: 0,
    substeps: [],
    parameters: {
      pipette: "right_pipette",
      pipette_obj: {
        name: "(Right) Single Channel 300 uL - Gen 1",
        category: "right_pipette",
      },
      pipette_strategy: "",
      volumes: [250],
      volumes_string: "250",
      source_sequence: "tube_rack",
      sourceSequenceId: 0,
      destination_sequence: "well_plate",
      destinationSequenceId: 1,
      liquidClass: "Default",
      liquidClassId: 0,
      tipsStrategy: "STANDARD",
      offset: {
        source: 1,
        destination: 1,
        source_type: "Bottom",
        destination_type: "Bottom",
      },
    },
  },
  {
    type: "loop",
    name: "Loop (1)",
    id: 1,
    num_iterations: 3,
    substeps: [
      {
        type: "simple_transfer",
        name: "Simple Transfer (2)",
        id: 2,
        substeps: [],
        parameters: {
          pipette: "right_pipette",
          pipette_obj: {
            name: "(Right) Single Channel 300 uL - Gen 1",
            category: "right_pipette",
          },
          pipette_strategy: "",
          volumes: [100],
          volumes_string: "100",
          source: { slot_number: 5, slot_name: "reservoir_5", wells: ["A1"] },
          destination: {
            slot_number: 1,
            slot_name: "wellplate_1",
            wells: ["D6"],
          },
          liquidClass: "Default",
          liquidClassId: 0,
          tipsStrategy: "ECO",
          offset: {
            source: 1,
            destination: 1,
            source_type: "Bottom",
            destination_type: "Bottom",
          },
        },
      },
    ],
  },
  {
    type: "pause",
    name: "Pause (3)",
    id: 3,
    substeps: [],
    auto_resume: false,
    pause_time: 15,
  },
  {
    type: "array_transfer",
    name: "Array Transfer (8)",
    id: 8,
    substeps: [],
    parameters: {
      pipette: "right_pipette",
      pipette_obj: {
        name: "(Right) Single Channel 300 uL - Gen 1",
        category: "right_pipette",
      },
      pipette_strategy: "",
      array_map_filename: "demo.csv",
      source_var_map: {
        "buffer 1": {
          slot_name: "tuberack_2",
          slot_number: 2,
          wells: ["A3"],
          destinations: {
            destination_plate: { volumes: ["500", "0"], wells: ["A1", "B1"] },
          },
        },
        "buffer 2": {
          slot_name: "tuberack_2",
          slot_number: 2,
          wells: ["A4"],
          destinations: {
            destination_plate: { volumes: ["0", "500"], wells: ["A1", "B1"] },
          },
        },
        water: {
          slot_name: "reservoir_5",
          slot_number: 5,
          wells: ["A1"],
          destinations: {
            destination_plate: {
              volumes: ["1000", "1000"],
              wells: ["A1", "B1"],
            },
          },
        },
      },
      destination_var_map: {
        destination_plate: { slot_name: "wellplate_1", slot_number: 1 },
      },
      hasMapConflict: false,
      mappingComplete: true,
      array_map: [
        ["tuberack_2", "A3", "wellplate_1", "A1", "500"],
        ["tuberack_2", "A3", "wellplate_1", "B1", "0"],
        ["tuberack_2", "A4", "wellplate_1", "A1", "0"],
        ["tuberack_2", "A4", "wellplate_1", "B1", "500"],
        ["reservoir_5", "A1", "wellplate_1", "A1", "1000"],
        ["reservoir_5", "A1", "wellplate_1", "B1", "1000"],
      ],
      liquidClass: "Default",
      liquidClassId: 0,
      tipsStrategy: "ECO",
      offset: {
        source: 1,
        destination: 1,
        source_type: "Bottom",
        destination_type: "Bottom",
      },
    },
  },
  {
    type: "bioshake_3000t",
    name: "Bioshake 3000T (4)",
    id: 4,
    substeps: [],
    parameters: {
      speed: 1500,
      setpoint: 25,
      tempControl: false,
      cooldown: true,
      duration: 10,
      wait_for_stop: true,
      force_stop: false,
    },
  },
  {
    type: "slack_message",
    name: "Slack Message (5)",
    id: 5,
    substeps: [],
    channel: "",
    webhook: "",
    message: "",
    channel_object: {},
  },
];
const sampleProtocolEditorVariables = {
  stepsstepID: -1,
  stepsFormType: "",
  chosenLiquidClassID: "",
  locSeqID: "",
  locSeqCategory: "",
};

describe("shallowMounted StepCardsView", () => {
  let getters;
  let store;
  let wrapper;

  beforeEach(() => {
    getters = {
      protocolEditorVariables: () => sampleProtocolEditorVariables,
    };

    store = new Vuex.Store({
      getters,
    });

    wrapper = shallowMount(StepCardsView, {
      store,
      localVue,
      propsData: {
        list: sampleList,
        value: sampleValue,
      },
    });
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
