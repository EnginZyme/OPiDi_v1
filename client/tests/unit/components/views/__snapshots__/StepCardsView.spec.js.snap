// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowMounted StepCardsView renders correctly 1`] = `
<draggable-stub
  animation="0"
  class="step_cards_view__step_cards_view"
  clone="[Function]"
  element="div"
  ghostclass="ghost"
  group="description"
  tag="div"
  value="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
>
  <div>
    <div
      class="step_cards_view__step_card"
    >
      <p
        class="step_cards_view__step_card__main_text"
      >
        Sequence Transfer (0)
      </p>
       
      <p
        class="step_cards_view__step_card__sub_text"
      >
        
        Sequence Transfer
      
      </p>
    </div>
     
    <!---->
  </div>
  <div>
    <div
      class="step_cards_view__step_card"
    >
      <p
        class="step_cards_view__step_card__main_text"
      >
        Loop (1)
      </p>
       
      <p
        class="step_cards_view__step_card__sub_text"
      >
        
        Loop
      
      </p>
    </div>
     
    <stepcardsview-stub
      class="step_cards_view__step_card__sub_step_cards_view"
      list="[object Object]"
    />
  </div>
  <div>
    <div
      class="step_cards_view__step_card"
    >
      <p
        class="step_cards_view__step_card__main_text"
      >
        Pause (3)
      </p>
       
      <p
        class="step_cards_view__step_card__sub_text"
      >
        
        Pause
      
      </p>
    </div>
     
    <!---->
  </div>
  <div>
    <div
      class="step_cards_view__step_card"
    >
      <p
        class="step_cards_view__step_card__main_text"
      >
        Array Transfer (8)
      </p>
       
      <p
        class="step_cards_view__step_card__sub_text"
      >
        
        Array Transfer
      
      </p>
    </div>
     
    <!---->
  </div>
  <div>
    <div
      class="step_cards_view__step_card"
    >
      <p
        class="step_cards_view__step_card__main_text"
      >
        Bioshake 3000T (4)
      </p>
       
      <p
        class="step_cards_view__step_card__sub_text"
      >
        
        Bioshake 3000T
      
      </p>
    </div>
     
    <!---->
  </div>
  <div>
    <div
      class="step_cards_view__step_card"
    >
      <p
        class="step_cards_view__step_card__main_text"
      >
        Slack Message (5)
      </p>
       
      <p
        class="step_cards_view__step_card__sub_text"
      >
        
        Slack Message
      
      </p>
    </div>
     
    <!---->
  </div>
</draggable-stub>
`;
